features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp

```diff
@@ -0,0 +1,582 @@
+                if (count == 0) {
+                    // Timeout with nothing received
+                    nsapi_error_size = NSAPI_ERROR_WOULD_BLOCK;
+                    success = false;
+                }
+                size = 0; // This simply to cause an exit
+            }
+        }
+    } else if (socket->proto == NSAPI_TCP) {
+        while (success && (size > 0)) {
+            read_blk = UBLOX_MAX_PACKET_SIZE;
+            if (read_blk > size) {
+                read_blk = size;
+            }
+            if (socket->pending_bytes > 0) {
+                _at.cmd_start("AT+USORD=");
+                _at.write_int(socket->id);
+                _at.write_int(read_blk);
+                _at.cmd_stop();
+
+                _at.resp_start("+USORD:");
+                _at.skip_param(); // receiving socket id
+                usorf_sz = _at.read_int();
+
+                // Must use what +USORD returns here as it may be less or more than we asked for
+                if (usorf_sz > socket->pending_bytes) {
+                    socket->pending_bytes = 0;
+                } else {
+                    socket->pending_bytes -= usorf_sz;
+                }
+
+                if (usorf_sz > size) {
+                    usorf_sz = size;
+                }
+                _at.read_bytes(&ch, 1);
+                _at.read_bytes((uint8_t *)buffer + count , usorf_sz);
+                _at.resp_stop();
+
+                if (usorf_sz > 0) {
+                    count += usorf_sz;
+                    size -= usorf_sz;
+                } else {
+                    success = false;
+                }
+            } else if (timer.read_ms() < SOCKET_TIMEOUT) {
+                // Wait for URCs
+                _at.process_oob();
+            } else {
+                if (count == 0) {
+                    // Timeout with nothing received
+                    nsapi_error_size = NSAPI_ERROR_WOULD_BLOCK;
+                    success = false;
+                }
+                size = 0; // This simply to cause an exit
+            }
+        }
+    }
+    timer.stop();
+    _at.unlock();
+
+
+    if (!count || (_at.get_last_error() != NSAPI_ERROR_OK)) {
+        return NSAPI_ERROR_WOULD_BLOCK;
+    } else {
+        nsapi_error_size = count;
+    }
+
+    if (success && socket->proto == NSAPI_UDP && address) {
+        address->set_ip_address(ipAddress);
+        address->get_ip_address();
+        address->set_port(port);
+    }
+
+    return nsapi_error_size;
+}
+
+nsapi_size_or_error_t UBLOX_AT_CellularStack::socket_sendto(nsapi_socket_t handle, const SocketAddress &addr, const void *data, unsigned size)
+{
+    CellularSocket *socket = (CellularSocket *)handle;
+    if (!socket || !socket->created) {
+        return NSAPI_ERROR_DEVICE_ERROR;
+    }
+
+    nsapi_size_or_error_t ret_val = NSAPI_ERROR_OK;
+
+    /* Check parameters */
+    if (addr.get_ip_version() == NSAPI_UNSPEC) {
+        return NSAPI_ERROR_DEVICE_ERROR;
+    }
+
+    _at.lock();
```

Comment:
`lock` not needed due to `socket_sendto_impl` has locks.

Reply:
Updated.